src/tokenizer.ts

```diff
@@ -3,98 +3,88 @@ import analyze from "./js.ts";
 
-        tokens.push([type, code, position]);
+      tokens.push([type, code, position]);
 
-        if (index === -1) {
-          break;
-        }
-
-        position += index;
-        source = source.slice(index);
-        type = source.startsWith("{{#") ? "comment" : "tag";
-        continue;
+      if (index === -1) {
+        break;
       }
 
-      if (type === "comment") {
-        source = source.slice(3);
-        const index = source.indexOf("#}}");
-        const comment = index === -1 ? source : source.slice(0, index);
-        tokens.push([type, comment, position]);
+      position += index;
+      source = source.slice(index);
+      type = source.startsWith("{{#") ? "comment" : "tag";
+      continue;
+    }
 
-        if (index === -1) {
-          break;
-        }
+    if (type === "comment") {
+      source = source.slice(3);
+      const index = source.indexOf("#}}");
+      const comment = index === -1 ? source : source.slice(0, index);
+      tokens.push([type, comment, position]);
 
-        position += index + 3;
-        source = source.slice(index + 3);
-        type = "string";
-        continue;
+      if (index === -1) {
+        break;
       }
 
-      if (type === "tag") {
-        const indexes = parseTag(source);
-        const lastIndex = indexes.length - 1;
-        let tag: Token | undefined;
+      position += index + 3;
+      source = source.slice(index + 3);
+      type = "string";
+      continue;
+    }
 
-        indexes.reduce((prev, curr, index) => {
-          const code = source.slice(prev, curr - 2);
+    if (type === "tag") {
+      const indexes = parseTag(source);
+      const lastIndex = indexes.length - 1;
+      let tag: Token | undefined;
 
-          // Tag
-          if (index === 1) {
-            tag = [type, code, position];
-            tokens.push(tag);
-            return curr;
-          }
+      indexes.reduce((prev, curr, index) => {
+        const code = source.slice(prev, curr - 2);
 
-          // Filters
-          tokens.push(["filter", code]);
+        // Tag
+        if (index === 1) {
+          tag = [type, code, position];
+          tokens.push(tag);
           return curr;
-        });
+        }
 
-        position += indexes[lastIndex];
-        source = source.slice(indexes[lastIndex]);
-        type = "string";
+        // Filters
+        tokens.push(["filter", code]);
+        return curr;
+      });
 
-        // Search the closing echo tag {{ /echo }}
-        if (tag?.[1].match(/^\-?\s*echo\s*\-?$/)) {
-          const end = source.match(/{{\-?\s*\/echo\s*\-?}}/);
+      position += indexes[lastIndex];
+      source = source.slice(indexes[lastIndex]);
+      type = "string";
 
-          if (!end) {
-            throw new Error("Unclosed echo tag");
-          }
+      // Search the closing echo tag {{ /echo }}
+      if (tag?.[1].match(/^\-?\s*echo\s*\-?$/)) {
+        const end = /{{\-?\s*\/echo\s*\-?}}/.exec(source);
```

Comment:
This is more of a TypeScript change; functionally the same, but TypeScript doesn't know that `end` will have an `index` property with `match` (it wouldn't if the regex had a `g` flag) but it knows with `exec`.

Reply:
Yep, that was already on my todo list for optimizations, just thought being consistent for now would be better :smile:

Reply:
understood!